Yield while building large results in Alternator - rjson::print, executor::batch_get_item

nuivall opened the pull request:
Adds preemption points used in Alternator when:
 - sending bigger json response
 - building results for BatchGetItem

I've tested manually by inserting in preemptible sections (e.g. before `os.write`) code similar to:

    auto start  = std::chrono::steady_clock::now();
    do { } while ((std::chrono::steady_clock::now() - start) < 100ms);

and seeing reactor stall times. After the patch they
were not increasing while before they kept building up due to no preemption.

Refs #7926
Fixes #13689

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3541/

avikivity commented on utils/rjson.cc:
```diff
@@ -251,7 +251,8 @@ future<> print(const rjson::value& value, seastar::output_stream<char>& os, size
                 _f = _f.then([this, buf = std::move(_buf), &os = _os, try_reuse]() mutable -> future<> {
-                    return os.write(buf.get(), buf.size()).then([this, buf = std::move(buf), try_reuse]() mutable {
+                    co_return co_await os.write(buf.get(), buf.size()).then([this, buf = std::move(buf), try_reuse]() mutable {
+
```
https://github.com/scylladb/seastar/blob/master/doc/tutorial.md#lambda-coroutines

Mixing coroutines and continuations is a recipe for confusion. Better to coroutinize everything.

avikivity replied:
I see the caller isn't a coroutine. Better to use non-coroutine yields then (seastar::maybe_yield()).

nyh replied:
By the way, does co_await automatically add preemption? If not, what in the above change adds preemption?
How did you test this works?

avikivity replied:
> By the way, does co_await automatically add preemption? 

Yes.

nyh commented on alternator/executor.cc:
```diff
@@ -3254,8 +3255,7 @@ future<executor::request_return_type> executor::batch_get_item(client_state& cli
                 utils::get_local_injector().inject("alternator_batch_get_item", [] { throw std::runtime_error("batch_get_item injection"); });
-                std::vector<rjson::value> jsons = describe_multi_item(schema, partition_slice, *selection, *qr.query_result, *attrs_to_get);
-                return make_ready_future<std::vector<rjson::value>>(std::move(jsons));
+                return describe_multi_item(std::move(schema), std::move(partition_slice), std::move(selection), std::move(qr.query_result), attrs_to_get);
```
why not std::move attrs_to_get the same as all the rest?

nuivall replied:
will add `move`

nyh reviewed (commented):
Thanks, looks good, I left only some minor comments. But please look at Avi's request.

avikivity commented on alternator/executor.cc:
```diff
@@ -2358,21 +2358,22 @@ std::optional<rjson::value> executor::describe_single_item(schema_ptr schema,
+        foreign_ptr<lw_shared_ptr<query::result>> query_result,
+        shared_ptr<const std::optional<attrs_to_get>> attrs_to_get) {
+    cql3::selection::result_set_builder builder(*selection, gc_clock::now(), cql_serialization_format::latest());
+    query::result_view::consume(*query_result, slice, cql3::selection::result_set_builder::visitor(builder, *schema, *selection));
```
While the result building non-preemptible time is now O(row), this consume() is still O(entire result).

nuivall replied:
That's true, how is this solved (or not) for cql ? I ran `pytest -vv -s test_paging.py::test_sticky_stop_flag` and see that the whole `SELECT` batch runs in a single loop inside `result_view::consume`.

I'm not sure how to easily change those cql visitors (mentioned also a similar case in https://github.com/scylladb/scylladb/pull/12351#issuecomment-1357388012). To my current understanding I'd need to wrap every function result (and all ancestor callers) in future<...>. Also taking extra care that there is no lambda involved.

It seems easy to do in alternator but I'd like to avoid touching cql for now as my code understanding is very limited there. Or am I missing some simpler way?

avikivity replied:
It's probably not solved. I don't even know if we need to solve it. For sure maybe-yielding everywhere will regress the common case when just a few rows are returned, so we'll have to special-case them.

nuivall commented:
v2:
- removed redundant flush
- rewrote streaming overload of `print()`
  - removed temporary_buffer reuse, it was making the code less readable without justification that it helps
  - I think it had use-after-move on `_buf`, when executing `_f.then([this, buf = std::move(_buf)` and then in continuation `if (try_reuse && _buf.empty()`
  

Additionally I ran `pytest -vv -s test_batch.py::test_batch_write_item_large` (with more queries in a loop) to see if there is no major perf regressions. Surprisingly json printing got 40% faster, but the only way I can explain it is that it's due to better cache utilization because similar code runs together and it probably won't hold in real workload (or maybe it will to some extend?). Screens before:
![Screenshot from 2023-01-20 18-20-58_before](https://user-images.githubusercontent.com/2939488/213772706-d0a27688-e993-4214-9021-b6c4a901f221.png)
after:
![Screenshot from 2023-01-20 18-20-27_after](https://user-images.githubusercontent.com/2939488/213772714-965a2c2d-7780-4e41-8ade-5f189e9361f1.png)

Also I've realized that it shouldn't be very difficult to add zero-copy writes to seastar's http_chunked_data_sink_impl which I'll try to address in a separate patch. Currently we're doing kind of double buffering.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3940/

nyh commented on utils/rjson.cc:
```diff
@@ -227,68 +227,60 @@ std::string print(const rjson::value& value, size_t max_nested_level) {
-                    });
-                });
-            }
+// This class implements rjson Handler and batches Put() calls into output_stream writes.
```
Also (I don't have where to write this so I'll do this here), please improve the commit message. Please explain that you are improving the print() function, and explain what you improved in it (I couldn't figure out how the new code is more "yieldable" than the previous one?)

Also, I'm not sure I ever understood the "reuse buffer" code we had here in the past, but @elcalio (I think?) added it for a reason, why is it no longer needed or beneficial?

nuivall replied:
> RapidJSON not "rjson"

will change

> please improve the commit message

will change

> reuse buffer

three reasons for removing:
 - doesn't seem to improve anything in the profile
 - once I add zero copy in seastar for chunked output stream we won't be able to reuse anyway I think
 - complicates the code (e.g. see use-after-move mention in https://github.com/scylladb/scylladb/pull/12351#issuecomment-1398751442)

nyh replied:
@elcallio can you please see if you remember why you did that "try_reuse" thing and what you saw it improved? It was a fairly complex piece of code, so I assume it served some purpose, I just don't remember what it was, and I can't find comments or commit messages explaining its purpose...

nuivall replied:
@elcallio ping, any objection removing reuse?

eliransin replied:
> @elcallio can you please see if you remember why you did that "try_reuse" thing and what you saw it improved?
@elcallio  can you give your input about this one?
Thanks

nyh reviewed (commented):
The first and third patch looks great, the second is less obvious to me so I had some questions there.

nyh commented:
>       * removed temporary_buffer reuse, it was making the code less readable without justification that it helps

Please explain this in the commit message. @elcallio you wrote this "try_reuse" code (commit e2d7225df8410de9f5ecdc4190f9cd14093b7ce0), maybe you can comment on what was the justification from it? If I ever knew, I forgot already :-(

A wild guess: Maybe the "try_reuse" code may be optimized for **small** prints, which may be important in other situations, not the large batch case.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3951/

nuivall commented on utils/rjson.cc:
```diff
@@ -227,68 +227,60 @@ std::string print(const rjson::value& value, size_t max_nested_level) {
+    static constexpr size_t buf_size = 512;
+    seastar::output_stream<char>& os;
+    temporary_buffer<char> buf = temporary_buffer<char>(buf_size);
+    size_t pos = 0;
```
it was accidental, missed the fact that private and public members follow different convention. will add `_`

nuivall commented:
> A wild guess: Maybe the "try_reuse" code may be optimized for **small** prints, which may be important in other situations, not the large batch case.

don't think so, it was actually unconditionally true (which was another small source of confusion to me) except the last write.

nuivall commented:
v3:
- improved commit message
- follow scylla naming conventions

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3969/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3968/

nuivall commented:
To increase confidence in this patch (in particular fairly complex change in json streaming) I added scan workload to https://github.com/scylladb/scylladb/pull/13121.

Test command:
```
❯ ninja ./build/release/scylla && ./build/release/scylla perf-alternator-workloads --workdir ~/tmp --smp 1 --developer-mode 1 --alternator-port 8000 --alternator-write-isolation only_rmw_uses_lwt --workload scan --duration 10 --scan-total-segments 10 2> /dev/null
```

Here are the measurements:

Yield (new):

```
88.21 tps (190387.5 allocs/op, 919.7 tasks/op, 98662410 insns/op,        0 errors)
89.73 tps (190711.0 allocs/op, 911.9 tasks/op, 98835105 insns/op,        0 errors)
87.34 tps (190912.8 allocs/op, 919.5 tasks/op, 98951341 insns/op,        0 errors)
88.15 tps (190435.3 allocs/op, 917.2 tasks/op, 98696135 insns/op,        0 errors)
89.35 tps (190070.7 allocs/op, 905.5 tasks/op, 98503717 insns/op,        0 errors)
89.85 tps (190180.1 allocs/op, 918.1 tasks/op, 98566685 insns/op,        0 errors)
89.60 tps (190245.9 allocs/op, 900.5 tasks/op, 98594531 insns/op,        0 errors)
89.74 tps (190582.9 allocs/op, 910.6 tasks/op, 98772719 insns/op,        0 errors)
89.46 tps (190740.7 allocs/op, 912.1 tasks/op, 98842016 insns/op,        0 errors)
88.98 tps (190442.5 allocs/op, 909.9 tasks/op, 98700738 insns/op,        0 errors)

median 89.45737590191733
median absolute deviation: 0.40
maximum: 89.85
minimum: 87.34
```

No yield (old):

```
89.90 tps (188160.7 allocs/op, 250.4 tasks/op, 99391626 insns/op,        0 errors)
87.30 tps (188455.1 allocs/op, 236.3 tasks/op, 99555856 insns/op,        0 errors)
89.13 tps (188711.2 allocs/op, 242.0 tasks/op, 99701488 insns/op,        0 errors)
89.98 tps (188281.3 allocs/op, 238.8 tasks/op, 99461387 insns/op,        0 errors)
90.25 tps (187751.8 allocs/op, 234.4 tasks/op, 99191405 insns/op,        0 errors)
87.69 tps (187890.9 allocs/op, 239.4 tasks/op, 99266825 insns/op,        0 errors)
88.67 tps (188049.1 allocs/op, 236.6 tasks/op, 99352365 insns/op,        0 errors)
89.45 tps (188409.4 allocs/op, 236.2 tasks/op, 99534313 insns/op,        0 errors)
85.56 tps (188354.8 allocs/op, 239.3 tasks/op, 99510375 insns/op,        0 errors)
89.69 tps (188332.7 allocs/op, 235.4 tasks/op, 99493719 insns/op,        0 errors)

median 88.93465277370424
median absolute deviation: 1.03
maximum: 90.28
minimum: 85.55
```

Summary (new/old):

```
allocs: 190740/188711  ~= 1.01
tasks: 912/242 ~= 3.77
insns: 98842016/99701488 ~= 0.99
```

Looks like after the patch allocations increase by 1%, tasks nr increase 3.8 times, instructions per op decreases by roughly 1%. I think all those values except task number are on the verge of statistical significance for those perf measurements.

Task increase can be explained as this is not the number of created tasks but it's being increased when task gets execution chunk, so due to more yielding tasks are switching more often.

-----------------------------------------

@nyh since you brought some question about small prints I tested plugging streaming code into GetItem and ran alternator perf tests on it (note that normally streaming code is not used for small prints).

Results:

```
baseline: 218849 insns/op
streaming (old code): 226821 insns/op
streaming and yielding (new code): 225863 insns/op
```

Summary: overhead of streaming small items is around 3.6% with old streaming code and 3.2% for new code.

-----------------------------------------

Bonus:
I was also wondering about buffer size, whether current 512 is optimal or not.

Here is different buf size comparison:

```
256 - 99M insns/op
512 - 98M insns/op
1024 - 98M insns/op
10240 - 98M insns/op
```

The 'exact' difference between 512 and 10240 is:
98754522/98335587 ~= 1.00426

So by increasing it we may gain only up to 0.4% tps (my testing scenario is streaming 400kB per req, while max supported by dynamo is 1MB). Perhaps worth increasing to 1024 to gain most of it but it's not crucial.

nuivall commented:
@nyh ping
